### PR TITLE
Refactor of the ledgerstate / Feat: Added Conflict models for BranchDAG

### PR DESCRIPTION
**Pull request by hmoog**
# Description of change

This PR adds models related to Conflicts in the BranchDAG.

## Type of change

- Enhancement (a non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Change checklist

- [x] My code follows the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: packages/ledgerstate/conflict.go [hunk 1]
```diff
@@ -0,0 +1,459 @@
...
+		err = xerrors.Errorf("failed to parse member count (%v): %w", err, ErrParseBytesFailed)
+		return
+	}
+	conflict.memberCount = int(memberCount)
+
+	return
+}
+
+// ConflictFromObjectStorage restores a Conflict object that was stored in the ObjectStorage.
+func ConflictFromObjectStorage(key []byte, data []byte) (outputMetadata *Conflict, err error) {
```
Comment (jonastheis) [inline packages/ledgerstate/conflict.go:225]:
```suggestion
func ConflictFromObjectStorage(key []byte, data []byte) (conflict *Conflict, err error) {
```

File: packages/ledgerstate/conflict.go [hunk 2]
```diff
@@ -0,0 +1,459 @@
...
+
+// ID returns the identifier of this Conflict.
+func (c *Conflict) ID() ConflictID {
+	return c.id
+}
+
+// MemberCount returns the amount of Branches that are part of this Conflict.
+func (c *Conflict) MemberCount() int {
+	c.memberCountMutex.RLock()
+	defer c.memberCountMutex.RLock()
```
Comment (jonastheis) [inline packages/ledgerstate/conflict.go:242]:
```suggestion
	defer c.memberCountMutex.RUnLock()
```

File: packages/ledgerstate/conflict.go [hunk 3]
```diff
@@ -0,0 +1,459 @@
...
+	if len(optionalDelta) >= 1 {
+		delta = optionalDelta[0]
+	}
+
+	c.memberCountMutex.Lock()
+	defer c.memberCountMutex.Unlock()
+
+	c.memberCount = c.memberCount - delta
+	c.SetModified()
+	newMemberCount = c.memberCount
```
Comment (jonastheis) [inline packages/ledgerstate/conflict.go:275]: Here you should be consistent in the way the method is structured. In `IncreaseMemberCount` you simply return `c.memberCount` vs here you introduce `newMemberCount`.

File: packages/ledgerstate/conflict.go [hunk 4]
```diff
@@ -0,0 +1,459 @@
...
+	panic("updates disabled")
+}
+
+// ObjectStorageKey returns the key that is used to store the object in the database. It is required to match the
+// StorableObject interface.
+func (c *Conflict) ObjectStorageKey() []byte {
+	return c.id.Bytes()
+}
+
+// ObjectStorageValue marshals the Output into a sequence of bytes. The ID is not serialized here as it is only used as
```
Comment (jonastheis) [inline packages/ledgerstate/conflict.go:304]:
```suggestion
// ObjectStorageValue marshals the Conflict into a sequence of bytes. The ID is not serialized here as it is only used as
```

File: packages/ledgerstate/conflict.go (hunk 3 above)
Comment (hmoog) [reply packages/ledgerstate/conflict.go:275]: okay I will add a named parameter as well as its a more dscriptive method signature then

File: packages/ledgerstate/conflict.go [hunk 5]
```diff
@@ -0,0 +1,458 @@
...
+		return
+	}
+	conflict.memberCount = int(memberCount)
+
+	return
+}
+
+// ConflictFromObjectStorage restores a Conflict object that was stored in the ObjectStorage.
+func ConflictFromObjectStorage(key []byte, data []byte) (conflict *Conflict, err error) {
+	if outputMetadata, _, err = ConflictFromBytes(byteutils.ConcatBytes(key, data)); err != nil {
```
Comment (jonastheis) [inline packages/ledgerstate/conflict.go:224]:
```suggestion
	if conflict, _, err = ConflictFromBytes(byteutils.ConcatBytes(key, data)); err != nil {
```

**Pull request merged**